dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/EnrollmentSMSListener.java

```diff
@@ -0,0 +1,122 @@
+package org.hisp.dhis.sms.listener;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.organisationunit.OrganisationUnitService;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstanceService;
+import org.hisp.dhis.program.ProgramService;
+import org.hisp.dhis.sms.incoming.IncomingSms;
+import org.hisp.dhis.sms.incoming.SmsMessageStatus;
+import org.hisp.dhis.sms.parse.SMSParserException;
+import org.hisp.dhis.smscompression.Consts;
+import org.hisp.dhis.smscompression.models.AttributeValue;
+import org.hisp.dhis.smscompression.models.EnrollmentSMSSubmission;
+import org.hisp.dhis.smscompression.models.SMSSubmission;
+import org.hisp.dhis.trackedentity.*;
+import org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValue;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Transactional
+public class EnrollmentSMSListener extends NewSMSListener {
+    @Autowired
+    private TrackedEntityInstanceService teiService;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private TrackedEntityTypeService trackedEntityTypeService;
+
+    @Autowired
+    private TrackedEntityAttributeService trackedEntityAttributeService;
+
+    @Autowired
+    private ProgramService programService;
+
+    @Autowired
+    private OrganisationUnitService organisationUnitService;
+
+    @Autowired
+    private ProgramInstanceService programInstanceService;
+
+    @Override
+    public boolean accept(IncomingSms sms) {
+        return sms != null;
+    }
+
+    @Override
+    protected void postProcess(IncomingSms sms, SMSSubmission submission) {
+        EnrollmentSMSSubmission subm = (EnrollmentSMSSubmission) submission;
+
+        Date date = subm.getTimestamp();
+        String sender = StringUtils.replace(sms.getOriginator(), "+", "");
+        Program program = programService.getProgram(subm.getTrackerProgram());
+        OrganisationUnit ou = organisationUnitService.getOrganisationUnit(subm.getOrgUnit());
+        TrackedEntityType entityType = trackedEntityTypeService.getTrackedEntityType(subm.getTrackedEntityType());
+        User user = userService.getUser(subm.getUserID());
+
+        if (!program.hasOrganisationUnit(ou)) {
+            sendFeedback(NO_OU_FOR_PROGRAM, sender, WARNING);
+            throw new SMSParserException(NO_OU_FOR_PROGRAM);
```

Comment:
This is an existing Exception? I'm not sure I'd consider it a parsing exception but handling all these errors as exceptions might be nice. Maybe we can create a new Exception. I hadn't decided how to handle it yet.